### PR TITLE
Fixed relative path traversal due to using version string in path

### PR DESCRIPTION
**Pull request by comrumino**
### Description
- It looks like rbenv addressed the [path traversal vulnerability some time ago.](https://github.com/rbenv/rbenv/commit/370c26a6c9ee0511972ea04904fcc89014a22987)
- There are some additional change available in rbenv that simplify the read logic. Even so, my changes are functionally the same.

### CODE REVIEW

File: libexec/pyenv-version-file-read [hunk 1]
```diff
@@ -11,9 +11,15 @@ if [ -s "$VERSION_FILE" ]; then
   IFS="${IFS}"$'\r'
   sep=
   while read -n 1024 -r version _ || [[ $version ]]; do
-      [[ -z $version || $version == \#* ]] && continue
-      printf "%s%s" "$sep" "$version"
-      sep=:
+    [[ -z $version || $version == \#* ]] && continue
+    # The version string is used to construct a path. Non-zero exit when
+    # invalid characters are found. This prevents issues such as CVE-2022-35861.
+    if [[ ! "$version" =~ "^[a-z0-9][-.0-9a-z]+$" ]]; then
```
Comment (native-api) [inline libexec/pyenv-version-file-read:17]: The Linux kernel allows any characters in a directory entry except `\0` and `/` (which is a path separator). So this looks overzealous to me. The Rbenv commit does it right.

**Review by native-api — COMMENTED** [review]
Do we really need to disallow path separators altogether?

AFAICS, the vulnerability rather lies in the fact that the resulting path can be outside of `$PYENV_ROOT/versions`. So it's technically sufficient to normalize the path and check if it lies within `$PYENV_ROOT/versions`.

That use case it not officially supported though -- so I wouldn't mind just disallowing path traversal constructs altogether like Rbenv did.

**Comment by native-api** [comment]
In the e-maiol, you mentioned having a CVE number assigned. Did you get it? If yes, it'll probably be worth mentioning.

**Comment by comrumino** [comment]
> In the e-mail, you mentioned having a CVE number assigned. Did you get it? If yes, it'll probably be worth mentioning.

I have CVE-2022-35861 reserved. 

> disallowing path traversal constructs altogether like Rbenv did.

Either approach makes sense to me. My original approach was more of an "allow" approach where rbenv took a "deny" approach.

**Comment by comrumino** [comment]
I am done amending my commit. Please review the newest commit.

The current CVE form fields are below. Let me know if you would like any revisions.

### Suggested description
pyenv 1.2.24 through 2.3.2
Allows local users to gain privileges via a .python-version file in the current working directory.
An attacker can craft a Python version string in .python-version
to execute shims under their control. (Shims are executables that
pass a command along to a specific version of pyenv. The version
string is used to construct the path to the command, and there is no
validation of whether the version specified is a valid version.)

### VulnerabilityType Other
Uncontrolled Search Path Element

### Vendor of Product
pyenv

### Affected Product Code Base
pyenv - 1.2.24-2.3.2

### Affected Component
.python-version,shims/*

### Attack Type
Context-dependent

### Impact Code execution
true

### Attack Vectors
Crafted .python-version file in the current directory allows attacker to specify shims under their control

### Reference
https://github.com/pyenv/pyenv
https://github.com/pyenv/pyenv/pull/2412

### Discoverer
James Stronz

File: libexec/pyenv-version-file-read [hunk 2]
```diff
@@ -6,16 +6,18 @@ set -e
...
+  # Read the first word from the specified version file. Avoid reading it whole.
   IFS="${IFS}"$'\r'
-  sep=
-  while read -n 1024 -r version _ || [[ $version ]]; do
-      [[ -z $version || $version == \#* ]] && continue
-      printf "%s%s" "$sep" "$version"
-      sep=:
-  done <"$VERSION_FILE"
-  [[ $sep ]] && { echo; exit; }
+  read -n 1024 -d "" -r version _ <"$VERSION_FILE" || :
```
Comment (native-api) [inline libexec/pyenv-version-file-read:11]: The removal of the loop will disallow parsing multiple versions in the file. Rbenv doesn't support that apparently.

Tests are also missing.

**Comment by native-api** [comment]
I tried to merge Rbenv codebase like @yyuu did before -- since there may be other useful changes besides this -- but was unsuccessful so far (some files didn't merge because of different names, perhaps they weren't properly renamed with Git when forking or merging before).

I'll continue trying, so whichever effort completes first will get into the codebase.

File: libexec/pyenv-version-file-read (hunk 2 above)
Comment (comrumino) [reply libexec/pyenv-version-file-read]: I amended my work again, this time I took the time to add some new unit tests as well :)

**Pull request merged**